File: server/lib/zatca.js
'use strict';
/**
 * محرك الفاتورة الإلكترونية (ZATCA) — المرحلة الأولى والثانية.
 *
 * المرحلة الأولى: توليد حمولة QR بترميز TLV ثم Base64 للحقول الخمسة الإلزامية.
 * المرحلة الثانية: إضافة الوسوم 6..9 (هاش الفاتورة SHA-256، التوقيع الرقمي ECDSA،
 *                 المفتاح العام، توقيع الشهادة) + سلسلة الفواتير (PIH) + UBL 2.1 XML.
 *
 * ملاحظة امتثال مهمة: الربط الفعلي والتوقيع المعتمد يتطلبان شهادة CSID صادرة من
 * هيئة الزكاة والضريبة والجمارك عبر واجهاتها. النظام يخزّن الشهادات والمفاتيح مشفّرة
 * ويستخدمها عند توفرها؛ وإذا لم تتوفر شهادة إنتاج يتم التوقيع بمفتاح محلي (وضع محاكاة)
 * ويُعلَّم ذلك بوضوح في الحقل signature_mode.
 */
const crypto = require('node:crypto');

// ---------------------------------------------------------------- TLV / QR

/** ترميز وسم واحد بصيغة Tag-Length-Value مع دعم القيم الطويلة. */
function tlv(tag, value) {
  const buf = Buffer.isBuffer(value) ? value : Buffer.from(String(value), 'utf8');
  if (buf.length <= 255) {
    return Buffer.concat([Buffer.from([tag, buf.length]), buf]);
  }
  // القيم الأطول من 255 بايت: ترميز الطول على بايتين (نمط ممتد)
  const len = Buffer.alloc(3);
  len[0] = 0x82; // long-form marker
  len.writeUInt16BE(buf.length, 1);
  return Buffer.concat([Buffer.from([tag]), len, buf]);
}

/**
 * بناء حمولة QR.
 * @param {object} p
 * @param {string} p.sellerName        اسم المنشأة المصدرة
 * @param {string} p.vatNumber         الرقم الضريبي للمنشأة
 * @param {string} p.timestamp         ISO 8601 (YYYY-MM-DDTHH:MM:SSZ)
 * @param {string} p.total             إجمالي الفاتورة شاملاً الضريبة (نص بمنزلتين)
 * @param {string} p.vatTotal          إجمالي الضريبة (نص بمنزلتين)
 * @param {string} [p.invoiceHash]     Tag 6 — هاش الفاتورة (base64)
 * @param {string} [p.signature]       Tag 7 — التوقيع الرقمي (base64)
 * @param {string} [p.publicKey]       Tag 8 — المفتاح العام (base64 DER)
 * @param {string} [p.certSignature]   Tag 9 — توقيع الشهادة (base64)
 * @returns {string} base64
 */
function buildQrPayload(p) {
  const parts = [
    tlv(1, p.sellerName || ''),
    tlv(2, p.vatNumber || ''),
    tlv(3, p.timestamp || ''),
    tlv(4, p.total || '0.00'),
    tlv(5, p.vatTotal || '0.00'),
  ];
  if (p.invoiceHash) parts.push(tlv(6, Buffer.from(p.invoiceHash, 'utf8')));
  if (p.signature) parts.push(tlv(7, Buffer.from(p.signature, 'base64')));
  if (p.publicKey) parts.push(tlv(8, Buffer.from(p.publicKey, 'base64')));
  if (p.certSignature) parts.push(tlv(9, Buffer.from(p.certSignature, 'base64')));
  return Buffer.concat(parts).toString('base64');
}

/** فك حمولة QR (يُستخدم في الاختبارات وشاشة التحقق). */
function parseQrPayload(base64) {
  const buf = Buffer.from(String(base64), 'base64');
  const out = [];
  let i = 0;
  while (i < buf.length) {
    const tag = buf[i];
    let len = buf[i + 1];
    let off = i + 2;
    if (len === 0x82) {
      len = buf.readUInt16BE(i + 2);
      off = i + 4;
    }
    const value = buf.subarray(off, off + len);
    out.push({ tag, length: len, value });
    i = off + len;
  }
  return out.map((t) => ({
    tag: t.tag,
    length: t.length,
    text: t.tag >= 7 ? t.value.toString('base64') : t.value.toString('utf8'),
  }));
}

// ---------------------------------------------------------------- Hashing / chaining

/** هاش الفاتورة: SHA-256 على تمثيل الفاتورة المُقنّن (canonical) ثم Base64. */
function invoiceHash(canonicalXml) {
  return crypto.createHash('sha256').update(Buffer.from(canonicalXml, 'utf8')).digest('base64');
}

/** قيمة PIH للفاتورة الأولى في السلسلة حسب مواصفات ZATCA. */
const GENESIS_PIH = Buffer.alloc(32, 0).toString('base64');

// ---------------------------------------------------------------- Signing

/** توليد زوج مفاتيح ECDSA (prime256v1) للاستخدام المحلي / طلب الشهادة. */
function generateKeyPair() {
  const { privateKey, publicKey } = crypto.generateKeyPairSync('ec', { namedCurve: 'prime256v1' });
  return {
    privateKeyPem: privateKey.export({ type: 'pkcs8', format: 'pem' }).toString(),
    publicKeyPem: publicKey.export({ type: 'spki', format: 'pem' }).toString(),
    publicKeyDerBase64: publicKey.export({ type: 'spki', format: 'der' }).toString('base64'),
  };
}

/** التوقيع الرقمي على هاش الفاتورة. يعيد base64. */
function signHash(privateKeyPem, hashBase64) {
  const key = crypto.createPrivateKey(privateKeyPem);
  const sig = crypto.sign('sha256', Buffer.from(hashBase64, 'utf8'), {
    key,
    dsaEncoding: 'der',
  });
  return sig.toString('base64');
}

function verifyHashSignature(publicKeyPem, hashBase64, signatureBase64) {
  try {
    return crypto.verify(
      'sha256',
      Buffer.from(hashBase64, 'utf8'),
      { key: crypto.createPublicKey(publicKeyPem), dsaEncoding: 'der' },
      Buffer.from(signatureBase64, 'base64'),
    );
  } catch {
    return false;
  }
}

/** استخراج المفتاح العام (DER base64) وتوقيع الشهادة من شهادة X.509 إن وُجدت. */
function certificateInfo(certPem) {
  try {
    const cert = new crypto.X509Certificate(certPem);
    return {
      publicKeyDerBase64: cert.publicKey.export({ type: 'spki', format: 'der' }).toString('base64'),
      serialNumber: cert.serialNumber,
      issuer: cert.issuer,
      subject: cert.subject,
      validTo: cert.validTo,
      // توقيع الشهادة: يُستخرج من بنية الشهادة (آخر عنصر في DER)
      certSignatureBase64: extractCertSignature(cert.raw),
    };
  } catch {
    return null;
  }
}

/** استخراج قيمة BIT STRING الأخيرة (signatureValue) من DER الشهادة. */
function extractCertSignature(der) {
  try {
    let i = der.length - 1;
    // البحث عن آخر وسم BIT STRING (0x03) بطول معقول
    for (let p = der.length - 80; p > 0; p--) {
      if (der[p] === 0x03 && der[p + 1] > 0x40 && der[p + 2] === 0x00) {
        const len = der[p + 1];
        return der.subarray(p + 3, p + 2 + len).toString('base64');
      }
    }
    void i;
    return null;
  } catch {
    return null;
  }
}

// ---------------------------------------------------------------- UBL 2.1 XML

function esc(value) {
  return String(value === null || value === undefined ? '' : value)
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&apos;');
}

/**
 * بناء ملف UBL 2.1 XML للفاتورة.
 * @param {object} inv   صف الفاتورة (بمبالغ عشرية جاهزة كنص)
 * @param {object} issuer بيانات المنشأة المصدرة
 * @param {object} client بيانات العميل
 * @param {Array}  lines  بنود الفاتورة
 * @param {object} chain  { icv, pih, currency }
 */
function buildUblXml(inv, issuer, client, lines, chain) {
  const currency = (chain && chain.currency) || 'SAR';
  const icv = (chain && chain.icv) || 1;
  const pih = (chain && chain.pih) || GENESIS_PIH;
  const typeCode = inv.invoice_type === 'SIMPLIFIED' ? '388' : '388';
  const typeName = inv.invoice_type === 'SIMPLIFIED' ? '0200000' : '0100000';

  const lineXml = lines
    .map((l, idx) => `
  <cac:InvoiceLine>
    <cbc:ID>${idx + 1}</cbc:ID>
    <cbc:InvoicedQuantity unitCode="${esc(l.unit || 'PCE')}">${esc(l.quantity)}</cbc:InvoicedQuantity>
    <cbc:LineExtensionAmount currencyID="${currency}">${esc(l.taxable)}</cbc:LineExtensionAmount>
    <cac:TaxTotal>
      <cbc:TaxAmount currencyID="${currency}">${esc(l.tax_amount)}</cbc:TaxAmount>
      <cbc:RoundingAmount currencyID="${currency}">${esc(l.total_line)}</cbc:RoundingAmount>
    </cac:TaxTotal>
    <cac:Item>
      <cbc:Name>${esc(l.item_name)}</cbc:Name>
      <cac:ClassifiedTaxCategory>
        <cbc:ID>S</cbc:ID>
        <cbc:Percent>${esc(l.tax_rate)}</cbc:Percent>
        <cac:TaxScheme><cbc:ID>VAT</cbc:ID></cac:TaxScheme>
      </cac:ClassifiedTaxCategory>
    </cac:Item>
    <cac:Price>
      <cbc:PriceAmount currencyID="${currency}">${esc(l.unit_price)}</cbc:PriceAmount>
    </cac:Price>
  </cac:InvoiceLine>`)
    .join('');

  return `<?xml version="1.0" encoding="UTF-8"?>
<Invoice xmlns="urn:oasis:names:specification:ubl:schema:xsd:Invoice-2"
  xmlns:cac="urn:oasis:names:specification:ubl:schema:xsd:CommonAggregateComponents-2"
  xmlns:cbc="urn:oasis:names:specification:ubl:schema:xsd:CommonBasicComponents-2"
  xmlns:ext="urn:oasis:names:specification:ubl:schema:xsd:CommonExtensionComponents-2">
  <cbc:ProfileID>reporting:1.0</cbc:ProfileID>
  <cbc:ID>${esc(inv.invoice_number)}</cbc:ID>
  <cbc:UUID>${esc(inv.uuid)}</cbc:UUID>
  <cbc:IssueDate>${esc(inv.issue_date)}</cbc:IssueDate>
  <cbc:IssueTime>${esc(inv.issue_time)}</cbc:IssueTime>
  <cbc:InvoiceTypeCode name="${typeName}">${typeCode}</cbc:InvoiceTypeCode>
  <cbc:DocumentCurrencyCode>${currency}</cbc:DocumentCurrencyCode>
  <cbc:TaxCurrencyCode>${currency}</cbc:TaxCurrencyCode>
  <cac:AdditionalDocumentReference>
    <cbc:ID>ICV</cbc:ID>
    <cbc:UUID>${esc(icv)}</cbc:UUID>
  </cac:AdditionalDocumentReference>
  <cac:AdditionalDocumentReference>
    <cbc:ID>PIH</cbc:ID>
    <cac:Attachment>
      <cbc:EmbeddedDocumentBinaryObject mimeCode="text/plain">${esc(pih)}</cbc:EmbeddedDocumentBinaryObject>
    </cac:Attachment>
  </cac:AdditionalDocumentReference>
  <cac:AccountingSupplierParty>
    <cac:Party>
      <cac:PartyIdentification>
        <cbc:ID schemeID="CRN">${esc(issuer.commercial_register)}</cbc:ID>
      </cac:PartyIdentification>
      <cac:PostalAddress>
        <cbc:StreetName>${esc(issuer.street)}</cbc:StreetName>
        <cbc:BuildingNumber>${esc(issuer.building_no)}</cbc:BuildingNumber>
        <cbc:CitySubdivisionName>${esc(issuer.district)}</cbc:CitySubdivisionName>
        <cbc:CityName>${esc(issuer.city)}</cbc:CityName>
        <cbc:PostalZone>${esc(issuer.postal_code)}</cbc:PostalZone>
        <cac:Country><cbc:IdentificationCode>${esc(issuer.country || 'SA')}</cbc:IdentificationCode></cac:Country>
      </cac:PostalAddress>
      <cac:PartyTaxScheme>
        <cbc:CompanyID>${esc(issuer.tax_number)}</cbc:CompanyID>
        <cac:TaxScheme><cbc:ID>VAT</cbc:ID></cac:TaxScheme>
      </cac:PartyTaxScheme>
      <cac:PartyLegalEntity>
        <cbc:RegistrationName>${esc(issuer.name_ar || issuer.name_en)}</cbc:RegistrationName>
      </cac:PartyLegalEntity>
    </cac:Party>
  </cac:AccountingSupplierParty>
  <cac:AccountingCustomerParty>
    <cac:Party>
      <cac:PostalAddress>
        <cbc:StreetName>${esc(client.address)}</cbc:StreetName>
        <cbc:CityName>${esc(client.city)}</cbc:CityName>
        <cac:Country><cbc:IdentificationCode>SA</cbc:IdentificationCode></cac:Country>
      </cac:PostalAddress>
      <cac:PartyTaxScheme>
        <cbc:CompanyID>${esc(client.tax_number)}</cbc:CompanyID>
        <cac:TaxScheme><cbc:ID>VAT</cbc:ID></cac:TaxScheme>
      </cac:PartyTaxScheme>
      <cac:PartyLegalEntity>
        <cbc:RegistrationName>${esc(client.name)}</cbc:RegistrationName>
      </cac:PartyLegalEntity>
    </cac:Party>
  </cac:AccountingCustomerParty>
  <cac:AllowanceCharge>
    <cbc:ChargeIndicator>false</cbc:ChargeIndicator>
    <cbc:AllowanceChargeReason>discount</cbc:AllowanceChargeReason>
    <cbc:Amount currencyID="${currency}">${esc(inv.discount_amount)}</cbc:Amount>
  </cac:AllowanceCharge>
  <cac:TaxTotal>
    <cbc:TaxAmount currencyID="${currency}">${esc(inv.tax_amount)}</cbc:TaxAmount>
    <cac:TaxSubtotal>
      <cbc:TaxableAmount currencyID="${currency}">${esc(inv.taxable_amount)}</cbc:TaxableAmount>
      <cbc:TaxAmount currencyID="${currency}">${esc(inv.tax_amount)}</cbc:TaxAmount>
      <cac:TaxCategory>
        <cbc:ID>S</cbc:ID>
        <cbc:Percent>${esc(inv.tax_rate_display)}</cbc:Percent>
        <cac:TaxScheme><cbc:ID>VAT</cbc:ID></cac:TaxScheme>
      </cac:TaxCategory>
    </cac:TaxSubtotal>
  </cac:TaxTotal>
  <cac:LegalMonetaryTotal>
    <cbc:LineExtensionAmount currencyID="${currency}">${esc(inv.subtotal)}</cbc:LineExtensionAmount>
    <cbc:TaxExclusiveAmount currencyID="${currency}">${esc(inv.taxable_amount)}</cbc:TaxExclusiveAmount>
    <cbc:TaxInclusiveAmount currencyID="${currency}">${esc(inv.grand_total)}</cbc:TaxInclusiveAmount>
    <cbc:AllowanceTotalAmount currencyID="${currency}">${esc(inv.discount_amount)}</cbc:AllowanceTotalAmount>
    <cbc:PayableAmount currencyID="${currency}">${esc(inv.grand_total)}</cbc:PayableAmount>
  </cac:LegalMonetaryTotal>${lineXml}
</Invoice>
`;
}

module.exports = {
  tlv,
  buildQrPayload,
  parseQrPayload,
  invoiceHash,
  GENESIS_PIH,
  generateKeyPair,
  signHash,
  verifyHashSignature,
  certificateInfo,
  buildUblXml,
};

File: internal/zatca/zatca.go
package zatca

import (
	"bytes"
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/binary"
	"encoding/pem"
	"fmt"
	"strings"
)

var GenesisPIH = base64.StdEncoding.EncodeToString(make([]byte, 32))

// TLV encodes a single Tag-Length-Value entry conforming to ZATCA specs.
func TLV(tag byte, val []byte) []byte {
	l := len(val)
	if l <= 255 {
		res := make([]byte, 2+l)
		res[0] = tag
		res[1] = byte(l)
		copy(res[2:], val)
		return res
	}
	// Extended length format (0x82 followed by 2 bytes BigEndian length)
	res := make([]byte, 4+l)
	res[0] = tag
	res[1] = 0x82
	binary.BigEndian.PutUint16(res[2:4], uint16(l))
	copy(res[4:], val)
	return res
}

type QrParams struct {
	SellerName    string
	VatNumber     string
	Timestamp     string
	Total         string
	VatTotal      string
	InvoiceHash   string
	Signature     string
	PublicKey     string
	CertSignature string
}

// BuildQrPayload constructs the TLV payload and returns it as a Base64 string.
func BuildQrPayload(p QrParams) string {
	var buf bytes.Buffer
	buf.Write(TLV(1, []byte(p.SellerName)))
	buf.Write(TLV(2, []byte(p.VatNumber)))
	buf.Write(TLV(3, []byte(p.Timestamp)))
	buf.Write(TLV(4, []byte(p.Total)))
	buf.Write(TLV(5, []byte(p.VatTotal)))

	if p.InvoiceHash != "" {
		buf.Write(TLV(6, []byte(p.InvoiceHash)))
	}
	if p.Signature != "" {
		if sigBytes, err := base64.StdEncoding.DecodeString(p.Signature); err == nil {
			buf.Write(TLV(7, sigBytes))
		}
	}
	if p.PublicKey != "" {
		if pkBytes, err := base64.StdEncoding.DecodeString(p.PublicKey); err == nil {
			buf.Write(TLV(8, pkBytes))
		}
	}
	if p.CertSignature != "" {
		if csBytes, err := base64.StdEncoding.DecodeString(p.CertSignature); err == nil {
			buf.Write(TLV(9, csBytes))
		}
	}

	return base64.StdEncoding.EncodeToString(buf.Bytes())
}

type QrTag struct {
	Tag    byte   `json:"tag"`
	Length int    `json:"length"`
	Text   string `json:"text"`
}

// ParseQrPayload decodes a base64 TLV QR string into parsed tags.
func ParseQrPayload(b64 string) ([]QrTag, error) {
	data, err := base64.StdEncoding.DecodeString(strings.TrimSpace(b64))
	if err != nil {
		return nil, err
	}

	var tags []QrTag
	i := 0
	for i < len(data) {
		tag := data[i]
		if i+1 >= len(data) {
			break
		}
		l := int(data[i+1])
		offset := i + 2
		if l == 0x82 {
			if i+3 >= len(data) {
				break
			}
			l = int(binary.BigEndian.Uint16(data[i+2 : i+4]))
			offset = i + 4
		}
		if offset+l > len(data) {
			break
		}
		val := data[offset : offset+l]
		text := ""
		if tag >= 7 {
			text = base64.StdEncoding.EncodeToString(val)
		} else {
			text = string(val)
		}
		tags = append(tags, QrTag{
			Tag:    tag,
			Length: l,
			Text:   text,
		})
		i = offset + l
	}
	return tags, nil
}

// InvoiceHash computes SHA-256 of XML then base64 encodes it.
func InvoiceHash(xmlStr string) string {
	h := sha256.Sum256([]byte(xmlStr))
	return base64.StdEncoding.EncodeToString(h[:])
}

type KeyPair struct {
	PrivateKeyPem      string
	PublicKeyPem       string
	PublicKeyDerBase64 string
}

// GenerateKeyPair generates a standard ECDSA P-256 (secp256r1) key pair.
func GenerateKeyPair() (*KeyPair, error) {
	priv, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	if err != nil {
		return nil, err
	}

	privBytes, err := x509.MarshalPKCS8PrivateKey(priv)
	if err != nil {
		return nil, err
	}
	privPem := pem.EncodeToMemory(&pem.Block{Type: "PRIVATE KEY", Bytes: privBytes})

	pubBytes, err := x509.MarshalPKIXPublicKey(&priv.PublicKey)
	if err != nil {
		return nil, err
	}
	pubPem := pem.EncodeToMemory(&pem.Block{Type: "PUBLIC KEY", Bytes: pubBytes})

	return &KeyPair{
		PrivateKeyPem:      string(privPem),
		PublicKeyPem:       string(pubPem),
		PublicKeyDerBase64: base64.StdEncoding.EncodeToString(pubBytes),
	}, nil
}

// SignHash signs the base64 invoice hash using ECDSA and returns ASN.1 DER in base64.
func SignHash(privateKeyPem, hashBase64 string) (string, error) {
	block, _ := pem.Decode([]byte(privateKeyPem))
	if block == nil {
		return "", fmt.Errorf("invalid private key PEM")
	}

	key, err := x509.ParsePKCS8PrivateKey(block.Bytes)
	if err != nil {
		// try EC private key
		key, err = x509.ParseECPrivateKey(block.Bytes)
		if err != nil {
			return "", err
		}
	}

	ecKey, ok := key.(*ecdsa.PrivateKey)
	if !ok {
		return "", fmt.Errorf("not an ECDSA private key")
	}

	digest := sha256.Sum256([]byte(hashBase64))
	sig, err := ecdsa.SignASN1(rand.Reader, ecKey, digest[:])
	if err != nil {
		return "", err
	}

	return base64.StdEncoding.EncodeToString(sig), nil
}

// VerifyHashSignature verifies the ASN.1 DER signature against hash and public key.
func VerifyHashSignature(publicKeyPem, hashBase64, signatureBase64 string) bool {
	block, _ := pem.Decode([]byte(publicKeyPem))
	if block == nil {
		return false
	}
	pub, err := x509.ParsePKIXPublicKey(block.Bytes)
	if err != nil {
		return false
	}
	ecPub, ok := pub.(*ecdsa.PublicKey)
	if !ok {
		return false
	}

	sig, err := base64.StdEncoding.DecodeString(signatureBase64)
	if err != nil {
		return false
	}

	digest := sha256.Sum256([]byte(hashBase64))
	return ecdsa.VerifyASN1(ecPub, digest[:], sig)
}

func escXml(s string) string {
	r := strings.ReplaceAll(s, "&", "&amp;")
	r = strings.ReplaceAll(r, "<", "&lt;")
	r = strings.ReplaceAll(r, ">", "&gt;")
	r = strings.ReplaceAll(r, "\"", "&quot;")
	r = strings.ReplaceAll(r, "'", "&apos;")
	return r
}

type UblInvoiceInfo struct {
	InvoiceNumber  string
	UUID           string
	IssueDate      string
	IssueTime      string
	InvoiceType    string
	Subtotal       string
	TaxableAmount  string
	TaxAmount      string
	GrandTotal     string
	DiscountAmount string
	TaxRateDisplay string
}

type UblPartyInfo struct {
	Name               string
	TaxNumber          string
	CommercialRegister string
	Street             string
	BuildingNo         string
	District           string
	City               string
	PostalCode         string
	Country            string
}

type UblLineItem struct {
	Name        string
	Unit        string
	Quantity    string
	UnitPrice   string
	TaxRate     string
	Taxable     string
	TaxAmount   string
	RoundingAmt string
}

type UblChainInfo struct {
	ICV      int64
	PIH      string
	Currency string
}

// BuildUblXml generates standard UBL 2.1 XML format for ZATCA e-invoices.
func BuildUblXml(inv UblInvoiceInfo, issuer UblPartyInfo, client UblPartyInfo, lines []UblLineItem, chain UblChainInfo) string {
	currency := chain.Currency
	if currency == "" {
		currency = "SAR"
	}
	pih := chain.PIH
	if pih == "" {
		pih = GenesisPIH
	}
	icv := chain.ICV
	if icv <= 0 {
		icv = 1
	}

	typeCode := "388"
	typeName := "0100000" // Standard Tax Invoice
	if inv.InvoiceType == "SIMPLIFIED" {
		typeName = "0200000" // Simplified Tax Invoice
	}

	var lineXml strings.Builder
	for idx, l := range lines {
		unit := l.Unit
		if unit == "" {
			unit = "PCE"
		}
		lineXml.WriteString(fmt.Sprintf(`
  <cac:InvoiceLine>
    <cbc:ID>%d</cbc:ID>
    <cbc:InvoicedQuantity unitCode="%s">%s</cbc:InvoicedQuantity>
    <cbc:LineExtensionAmount currencyID="%s">%s</cbc:LineExtensionAmount>
    <cac:TaxTotal>
      <cbc:TaxAmount currencyID="%s">%s</cbc:TaxAmount>
      <cbc:RoundingAmount currencyID="%s">%s</cbc:RoundingAmount>
    </cac:TaxTotal>
    <cac:Item>
      <cbc:Name>%s</cbc:Name>
      <cac:ClassifiedTaxCategory>
        <cbc:ID>S</cbc:ID>
        <cbc:Percent>%s</cbc:Percent>
        <cac:TaxScheme><cbc:ID>VAT</cbc:ID></cac:TaxScheme>
      </cac:ClassifiedTaxCategory>
    </cac:Item>
    <cac:Price>
      <cbc:PriceAmount currencyID="%s">%s</cbc:PriceAmount>
    </cac:Price>
  </cac:InvoiceLine>`, idx+1, escXml(unit), escXml(l.Quantity), currency, escXml(l.Taxable),
			currency, escXml(l.TaxAmount), currency, escXml(l.RoundingAmt),
			escXml(l.Name), escXml(l.TaxRate), currency, escXml(l.UnitPrice)))
	}

	country := issuer.Country
	if country == "" {
		country = "SA"
	}

	return fmt.Sprintf(`<?xml version="1.0" encoding="UTF-8"?>
<Invoice xmlns="urn:oasis:names:specification:ubl:schema:xsd:Invoice-2"
  xmlns:cac="urn:oasis:names:specification:ubl:schema:xsd:CommonAggregateComponents-2"
  xmlns:cbc="urn:oasis:names:specification:ubl:schema:xsd:CommonBasicComponents-2"
  xmlns:ext="urn:oasis:names:specification:ubl:schema:xsd:CommonExtensionComponents-2">
  <cbc:ProfileID>reporting:1.0</cbc:ProfileID>
  <cbc:ID>%s</cbc:ID>
  <cbc:UUID>%s</cbc:UUID>
  <cbc:IssueDate>%s</cbc:IssueDate>
  <cbc:IssueTime>%s</cbc:IssueTime>
  <cbc:InvoiceTypeCode name="%s">%s</cbc:InvoiceTypeCode>
  <cbc:DocumentCurrencyCode>%s</cbc:DocumentCurrencyCode>
  <cbc:TaxCurrencyCode>%s</cbc:TaxCurrencyCode>
  <cac:AdditionalDocumentReference>
    <cbc:ID>ICV</cbc:ID>
    <cbc:UUID>%d</cbc:UUID>
  </cac:AdditionalDocumentReference>
  <cac:AdditionalDocumentReference>
    <cbc:ID>PIH</cbc:ID>
    <cac:Attachment>
      <cbc:EmbeddedDocumentBinaryObject mimeCode="text/plain">%s</cbc:EmbeddedDocumentBinaryObject>
    </cac:Attachment>
  </cac:AdditionalDocumentReference>
  <cac:AccountingSupplierParty>
    <cac:Party>
      <cac:PartyIdentification>
        <cbc:ID schemeID="CRN">%s</cbc:ID>
      </cac:PartyIdentification>
      <cac:PostalAddress>
        <cbc:StreetName>%s</cbc:StreetName>
        <cbc:BuildingNumber>%s</cbc:BuildingNumber>
        <cbc:CitySubdivisionName>%s</cbc:CitySubdivisionName>
        <cbc:CityName>%s</cbc:CityName>
        <cbc:PostalZone>%s</cbc:PostalZone>
        <cac:Country><cbc:IdentificationCode>%s</cbc:IdentificationCode></cac:Country>
      </cac:PostalAddress>
      <cac:PartyTaxScheme>
        <cbc:CompanyID>%s</cbc:CompanyID>
        <cac:TaxScheme><cbc:ID>VAT</cbc:ID></cac:TaxScheme>
      </cac:PartyTaxScheme>
      <cac:PartyLegalEntity>
        <cbc:RegistrationName>%s</cbc:RegistrationName>
      </cac:PartyLegalEntity>
    </cac:Party>
  </cac:AccountingSupplierParty>
  <cac:AccountingCustomerParty>
    <cac:Party>
      <cac:PostalAddress>
        <cbc:StreetName>%s</cbc:StreetName>
        <cbc:CityName>%s</cbc:CityName>
        <cac:Country><cbc:IdentificationCode>SA</cbc:IdentificationCode></cac:Country>
      </cac:PostalAddress>
      <cac:PartyTaxScheme>
        <cbc:CompanyID>%s</cbc:CompanyID>
        <cac:TaxScheme><cbc:ID>VAT</cbc:ID></cac:TaxScheme>
      </cac:PartyTaxScheme>
      <cac:PartyLegalEntity>
        <cbc:RegistrationName>%s</cbc:RegistrationName>
      </cac:PartyLegalEntity>
    </cac:Party>
  </cac:AccountingCustomerParty>
  <cac:AllowanceCharge>
    <cbc:ChargeIndicator>false</cbc:ChargeIndicator>
    <cbc:AllowanceChargeReason>discount</cbc:AllowanceChargeReason>
    <cbc:Amount currencyID="%s">%s</cbc:Amount>
  </cac:AllowanceCharge>
  <cac:TaxTotal>
    <cbc:TaxAmount currencyID="%s">%s</cbc:TaxAmount>
    <cac:TaxSubtotal>
      <cbc:TaxableAmount currencyID="%s">%s</cbc:TaxableAmount>
      <cbc:TaxAmount currencyID="%s">%s</cbc:TaxAmount>
      <cac:TaxCategory>
        <cbc:ID>S</cbc:ID>
        <cbc:Percent>%s</cbc:Percent>
        <cac:TaxScheme><cbc:ID>VAT</cbc:ID></cac:TaxScheme>
      </cac:TaxCategory>
    </cac:TaxSubtotal>
  </cac:TaxTotal>
  <cac:LegalMonetaryTotal>
    <cbc:LineExtensionAmount currencyID="%s">%s</cbc:LineExtensionAmount>
    <cbc:TaxExclusiveAmount currencyID="%s">%s</cbc:TaxExclusiveAmount>
    <cbc:TaxInclusiveAmount currencyID="%s">%s</cbc:TaxInclusiveAmount>
    <cbc:AllowanceTotalAmount currencyID="%s">%s</cbc:AllowanceTotalAmount>
    <cbc:PayableAmount currencyID="%s">%s</cbc:PayableAmount>
  </cac:LegalMonetaryTotal>%s
</Invoice>
`, escXml(inv.InvoiceNumber), escXml(inv.UUID), escXml(inv.IssueDate), escXml(inv.IssueTime),
		typeName, typeCode, currency, currency, icv, escXml(pih),
		escXml(issuer.CommercialRegister), escXml(issuer.Street), escXml(issuer.BuildingNo),
		escXml(issuer.District), escXml(issuer.City), escXml(issuer.PostalCode), escXml(country),
		escXml(issuer.TaxNumber), escXml(issuer.Name),
		escXml(client.Street), escXml(client.City), escXml(client.TaxNumber), escXml(client.Name),
		currency, escXml(inv.DiscountAmount),
		currency, escXml(inv.TaxAmount),
		currency, escXml(inv.TaxableAmount), currency, escXml(inv.TaxAmount), escXml(inv.TaxRateDisplay),
		currency, escXml(inv.Subtotal), currency, escXml(inv.TaxableAmount), currency, escXml(inv.GrandTotal),
		currency, escXml(inv.DiscountAmount), currency, escXml(inv.GrandTotal), lineXml.String())
}
